src/parser/btype.rs

```diff
@@ -0,0 +1,120 @@
+use crate::helpers::variable_bit_structures::VarBitInt;
+use crate::instructions::ParsedInstruction32;
+use crate::registers::Register;
+
+pub(crate) fn parse_btype32(_opcode: &u8, imm: &VarBitInt, funct3: &u8, rs1: &u8, rs2: &u8) -> Result<ParsedInstruction32, &'static str> {
+    match funct3 {
+        0b000 => Ok(ParsedInstruction32::beq {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        0b001 => Ok(ParsedInstruction32::bne {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        0b100 => Ok(ParsedInstruction32::blt {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        0b101 => Ok(ParsedInstruction32::bge {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        0b110 => Ok(ParsedInstruction32::bltu {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        0b111 => Ok(ParsedInstruction32::bgeu {
+            rs1: Register::try_from(*rs1)?,
+            rs2: Register::try_from(*rs2)?,
+            imm: i32::try_from(*imm)?,
+        }),
+        _ => Err("Invalid funct3"),
+    }   
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::instructions::ParsedInstruction32;
+    use crate::registers::Register;
+
+    #[test]
+    fn test_parse_btype32_beq() {
+        let imm = VarBitInt::new(0b000000000001, 12);
+        let result = parse_btype32(&0b1100011, &imm, &0b000, &0b00010, &0b00011).unwrap();
+        assert_eq!(result, ParsedInstruction32::beq {
+            rs1: Register::try_from(0b00010).unwrap(),
+            rs2: Register::try_from(0b00011).unwrap(),
+            imm: i32::try_from(imm).unwrap(),
+        });
+    }
+
+    #[test]
+    fn test_parse_btype32_bne() {
+        let imm = VarBitInt::new(0b000000000001, 12);
```

Comment:
The only thing being tested here is the correct instruction being parsed (as the rest of the test is just a copy of the program code). I believe it would be simpler and clearer to do something like:

```suggestion
         match parse_btype32(&0b1100011, &imm, &0b000, &0b00010, &0b00011).unwrap() {
            ParsedInstruction32::beq { .. } => {
                assert!(true);
            }
            _ => panic!("beq instruction decoded as a different type"),
        }
```

Reply:
I refactored the unit tests to only test for correct enum, also refactored the parsing to improve readability